### PR TITLE
Add Withdrawal Pattern Example

### LINKED ISSUE #705: Withdrawing Pattern Example
Currently, in the security considerations recommendations, there is a suggestion of a pattern to use withdrawal, but there is no documentation of such a pattern in the "common patterns" section. If someone could please put that in, that would be greatly appreciated.

### PR DESCRIPTION
**Pull request by Denton-L**
I feel like this is very rough, currently. Please let me know about what improvements I can make with this!

resolve #705

### CODE REVIEW

**Comment by Denton-L** [comment]
I have no idea why the build is failing right now.

**Comment by VoR0220** [comment]
glitch in the matrix I suppose

File: docs/common-patterns.rst [hunk 1]
```diff
@@ -2,6 +2,82 @@
...
+.. index:: withdrawal
+
+.. _withdrawal_pattern:
+
+*************************
+Withdrawal from Contracts
+*************************
+
+The recommended method of sending funds after an effect
+is with the withdrawal pattern. Although the most intuitive
```
Comment (chriseth) [inline docs/common-patterns.rst]: `is using the`?

File: docs/common-patterns.rst [hunk 2]
```diff
@@ -2,6 +2,82 @@
...
+aethod of sending Ether as a result of an effect is a
```
Comment (chriseth) [inline docs/common-patterns.rst]: `method`

File: docs/common-patterns.rst [hunk 3]
```diff
@@ -2,6 +2,82 @@
...
+                throw;
+            }
+        }
+
+        function () {
+            etherStore[msg.sender] += msg.value;
+        }
+    }
+
+This is as opposed to the more intuitive sending pattern.
```
Comment (chriseth) [inline docs/common-patterns.rst:69]: We could add that it is fine in this extremely simple example, but gets harder and harder to handle correctly with more complex contracts.

File: docs/common-patterns.rst [hunk 4]
```diff
@@ -2,6 +2,82 @@
...
+
+::
+
+    contract WithdrawalPattern {
+
+        mapping (address => uint) etherStore;
+        mapping (address => uint) pendingReturns;
+
+        function sendEther(uint amount) {
+            if (amount < etherStore[msg.sender]) {
```
Comment (chriseth) [inline docs/common-patterns.rst]: `amount > etherStore[...`

File: docs/common-patterns.rst [hunk 5]
```diff
@@ -2,6 +2,82 @@
...
+
+This is as opposed to the more intuitive sending pattern.
+
+::
+
+    contract SendPattern {
+
+        mapping (address => uint) etherStore;
+
+        function sendEther(uint amount) {
```
Comment (chriseth) [inline docs/common-patterns.rst]: This is basically also the withdraw pattern. I think we should create a more complicated example where we have an actual effect.

**Comment by Denton-L** [comment]
@chriseth, do you have any suggestions on what kind of example might be a good one to use?

**Comment by chriseth** [comment]
@Denton-L the auction example already has the withdrawal pattern, perhaps we could borrow a snippet from there?

**Comment by Denton-L** [comment]
@chriseth, I have included a modified version of the auction example.

**Comment by chriseth** [comment]
Sorry, but I'm not really keen on just repeating the auction example here. If the pattern cannot be shown properly in isolation, then we should rather link to the auction example.

**Comment by Denton-L** [comment]
Sorry, I've been busy for the past little bit. I've updated it so that it uses a different example, the "store" example. This should be shorter and simpler than the auction example.

**Comment by chriseth** [comment]
Thanks for the update. Sorry to say that, but I think that also this is not the best example. Examples where you really need the withdrawal patterns are those where the person triggering the action does not necessarily benefit from the transfer.

In the example here, you could just use `if(!x.send(y)) throw` - and in fact you should because otherwise this example will generate a compiler warning. Using `throw` does not harm here, because it does not stall a process. Where you cannot use throw and have to use withdraw is where the sending is part of another process that can then be stalled by the person receiving the funds.

**Comment by Denton-L** [comment]
Would a [King of the Ether](https://www.kingoftheether.com/)-type example work, then?

It acts as almost an Auction contract without as much complexity.

**Comment by chriseth** [comment]
Yes, that's a good idea!

File: docs/common-patterns.rst [hunk 6]
```diff
@@ -2,6 +2,105 @@
...
+*************************
+
+The recommended method of sending funds after an effect
+is using the withdrawal pattern. Although the most intuitive
+method of sending Ether, as a result of an effect, is a
+direct ``send`` call, this is not recommended as it
+introduces a potential security risk. You may read
+more about this on the :ref:`security_considerations` page.
+
+This is an example of the withdrawal pattern in practice in
```
Comment (chriseth) [inline docs/common-patterns.rst:20]: I think we should mention "king of the ether" as insipration here.

File: docs/common-patterns.rst [hunk 7]
```diff
@@ -2,6 +2,105 @@
...
+you will recieve the funds of the person who has gone on to
+become the new richest.
+
+::
+
+    contract WithdrawalContract {
+        address public richest;
+        uint public mostSent;
+
+        mapping (address => uint) pending;
```
Comment (chriseth) [inline docs/common-patterns.rst]: `pendingWithdrawal`?

File: docs/common-patterns.rst [hunk 8]
```diff
@@ -2,6 +2,105 @@
...
+
+        function WithdrawalContract() {
+            richest = msg.sender;
+            mostSent = msg.value;
+        }
+
+        function becomeRichest() returns (bool) {
+            if (msg.value > mostSent) {
+                pending[richest] = msg.value;
```
Comment (chriseth) [inline docs/common-patterns.rst]: I think we need `pending[richest] += msg.value` here.

**Comment by Denton-L** [comment]
@chriseth, as in #913, I changed the example such that it does not throw anymore.

File: docs/common-patterns.rst [hunk 9]
```diff
@@ -2,6 +2,110 @@
...
+        mapping (address => uint) pendingWithdrawals;
+
+        function WithdrawalContract() {
+            richest = msg.sender;
+            mostSent = msg.value;
+        }
+
+        function becomeRichest() returns (bool) {
+            if (msg.value > mostSent) {
+                pendingWithdrawals[richest] += msg.value;
```
Comment (chriseth) [inline docs/common-patterns.rst:44]: I think it should be `pendingWithdrawals[richest] += mostSent` here.

File: docs/common-patterns.rst [hunk 10]
```diff
@@ -2,6 +2,110 @@
...
+            richest = msg.sender;
+            mostSent = msg.value;
+        }
+
+        function becomeRichest() returns (bool) {
+            if (msg.value > mostSent) {
+                // Check if call succeeds to prevent an attacker
+                // from trapping the previous person's funds in
+                // this contract through a callstack attack
+                if (!richest.send(msg.value)) {
```
Comment (chriseth) [inline docs/common-patterns.rst:87]: `richest.send(mostSent)`

File: docs/common-patterns.rst (hunk 9 above)
Comment (Denton-L) [inline docs/common-patterns.rst:44]: I think the point of this is that your profit is the difference of whatever the person after sent in. That way, each person makes a little bit of money except for the very last person to send in their Ether.

File: docs/common-patterns.rst (hunk 10 above)
Comment (Denton-L) [inline docs/common-patterns.rst:87]: See other comment.

File: docs/common-patterns.rst (hunk 9 above)
Comment (chriseth) [inline docs/common-patterns.rst:44]: Ah, it seems I completely misunderstood that example :-)

**Pull request merged**